extension/json/json_common.cpp

```diff
@@ -31,24 +32,65 @@ string ThrowPathError(const char *ptr, const char *end, const bool binder) {
 	}
 }
 
-static inline idx_t ReadString(const char *ptr, const char *const end, const bool escaped) {
+struct JSONKeyReadResult {
+public:
+	static inline JSONKeyReadResult Empty() {
+		return {idx_t(0), string()};
+	}
+
+	static inline JSONKeyReadResult WildCard() {
+		return {1, "*"};
+	}
+
+	inline bool IsValid() {
+		return chars_read != 0;
+	}
+
+	inline bool IsWildCard() {
+		return key == "*";
+	}
+
+public:
+	idx_t chars_read;
+	string key;
+};
+
+static inline JSONKeyReadResult ReadString(const char *ptr, const char *const end, const bool escaped) {
 	const char *const before = ptr;
 	if (escaped) {
+		auto key = make_unsafe_uniq_array<char>(end - ptr);
+		idx_t key_len = 0;
+
+		bool backslash = false;
 		while (ptr != end) {
 			if (*ptr == '"') {
-				break;
+				if (!backslash) {
+					break;
+				}
+				backslash = false;
+			} else if (*ptr == '\\') {
+				if (!backslash) {
+					backslash = true;
+					ptr++;
+					continue;
+				}
+				backslash = false;
 			}
-			ptr++;
+			key[key_len++] = *ptr++;
+		}
+		if (ptr == end || backslash) {
```

Comment:
Should this not be an error if `backslash` is true?

Reply:
we ignore other backslashes as we do for regular (non quoted) JSON paths

Reply:
Hm, but then returning an empty result here doesn't seem correct to me - shouldn't we match `x\` as the string literal `'x\'` in that case, instead of returning an empty result? Could we add such a test?

Reply:
Ah right, I understand the confusion. If `ptr == end || backslash`, we return an empty result, which will cause a descriptive error, pointing to the location where the error occurred. So, if we have `$."x\` as a path, you will get an error

Reply:
That seems a bit inconsistent, no? We accept a backslash as a regular backslash except when it is the last character? I would say we should turn it into a regular backslash in this scenario perhaps?

Reply:
When you have the path `$."x\`, we need to throw an error, because the path is "escaped", i.e., the object key starts with a `"`, so it needs to end with a `"` too. If it ends with a backslash, it's wrong. Actually, I just realised that it can be removed from the check, since, `ptr == end` will be true